django/thunderstore/api/cyberstorm/views/packages.py

```diff
@@ -0,0 +1,467 @@
+
+    def _paginate(self, params: OrderedDict, queryset: QuerySet[Package]) -> Page:
+        """
+        Slice queryset based on the requested page.
+        """
+        paginator = CachedPaginator(
+            queryset,
+            self.page_size,
+            cache_key=self._get_paginator_cache_key(),
+            cache_vary=self._get_full_cache_vary(params),
+            cache_bust_condition=CacheBustCondition.any_package_updated,
+        )
+
+        # PageNotAnInteger error won't be raised here since deserializer
+        # has already checked that the page parameter is an integer.
+        try:
+            page = paginator.page(params["page"])
+        except EmptyPage:
+            raise ParseError("Page index error: no results on requested page")
+
+        return page
+
+    def _serialize_results(
+        self,
+        params: OrderedDict,
+        package_page: Page,
+    ) -> PackageListResponseSerializer:
+        """
+        Format results to transportation.
+        """
+        packages = [
+            {
+                "categories": p.community_listings.all()[0].categories.all(),
+                "community_identifier": p.community_listings.all()[
+                    0
+                ].community.identifier,
+                "description": p.latest.description,
+                "download_count": p.download_count,
+                "icon_url": p.latest.icon.url if bool(p.latest.icon) else None,
+                "is_deprecated": p.is_deprecated,
+                "is_nsfw": p.community_listings.all()[0].has_nsfw_content,
+                "is_pinned": p.is_pinned,
+                "last_updated": p.date_updated,
+                "namespace": p.namespace.name,
+                "name": p.name,
+                "rating_count": p.rating_count,
+                "size": p.latest.file_size,
+            }
+            for p in package_page.object_list
+        ]
+
+        (prev_url, next_url) = self._get_sibling_pages(params, package_page)
+
+        return PackageListResponseSerializer(
+            {
+                "count": package_page.paginator.count,
+                "previous": prev_url,
+                "next": next_url,
+                "results": packages,
+            },
+        )
+
+    def _get_full_cache_vary(self, params: OrderedDict) -> str:
+        """
+        Return cache key for CachedPaginator.
+        """
+        cache_vary = self._get_paginator_cache_vary_prefix()
+        cache_vary += f".{params['deprecated']}"
+        cache_vary += f".{params['nsfw']}"
+        cache_vary += f".{sorted(params['included_categories'])}"
+        cache_vary += f".{sorted(params['excluded_categories'])}"
+        cache_vary += f".{params.get('section', '-')}"
+        cache_vary += f".{params.get('q', '-')}"
+        cache_vary += f".{params['ordering']}"
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{self.page_size}"
+        return cache_vary
```

Comment:
Since we're already dealing with serializers, I suspect this could be simplified by rendering it through a serializer (which has consistent field ordering) & hashing or b64 encoding the result. Sorting the selected categories might not be possible as easily with that approach however.

Main reason for suggesting is to more concretely tie the cache key with the parameters the view was actually called with, hopefully avoiding bugs where one part of the code is changed in the future (e.g. new parameter introduced) without updating the cache key vary here. Technically it would suffice to use the query string directly, but that might not be consistently ordered & thus lead to duplicates, which is the main reason it was manually defined in the Django view implementation.

Not a blocker but open for improvement ideas

Reply:
Not sure what was meant by "rendering it through a serializer"? Serializing the OrderedDict returned by the DRF's serializer class (i.e. `params` here)? Anyway, didn't do any changes related to this at this time.

Reply:
That was referring to how the request parameters are validated with a serializer instance already, which also means we could use the same serializer instance to render them back to JSON for example (which should automatically order the fields as they're defined in the serializer & strip extra data if any was provided).

In practice I don't think it's a good idea to require cache key generation to be the responsibility of each view individually. In the case of (new) API views, we can cache the entire response & the factors we'd want to vary it on are:
- Request path
- Request parameters

We'd also want to ensure the cache key is generated in a consistent fashion (i.e. ordering & deduplication) and any invalid extra parameters are dropped (which happens now that it goes through a serializer for validation).

None of the requirements laid out so far require view specific implementation (outside of parameter validation to drop the excess parameters perhaps), which was the premise of my initial comment. Implementing a solution that'd automatically build the cache keys for API views via a middleware or a mixin should reduce the likelihood of caching bugs due to human error in the future. Caching bugs are one of the more common bug classes in the history of the project, thus the focus afforded to them here.

I do think implementing a generic caching solution is out of scope for this specific PR, but I'm slightly leaning towards excluding caching logic from this PR entirely as to not promote a solution that shouldn't be used. django-cachalot should be handling queryset caching implicitly already so the performance is _probably_ fine for the time being; comes down to the JSON rendering speed.

As for the question "why does this code exist in the legacy views if it's bad", the main reason is due to how the templates are partially cached in blocks rather than the entire response being cached, which led to this paradigm. In retrospect, even the legacy frontend views could utilize a generic mixin for building the cache key automatically rather than requiring the views to figure it out themselves, but that didn't cross my mind at the time due to focus on template caching.

**In summary, your options are roughly:**
1. Tell me that the caching is necessary for the performance of this view & we should stick with it regardless of what I laid out above, make no changes to it.
2. Remove the caching logic, replace it with nothing. We'll sort out a more generic (e.g. mixin based) caching solution for API views at a later date. Irrespective of what happens with this PR, we still need to work on a generic caching solution if we want to leverage CDN-based caching for the API, meaning this option can be considered "free". The risk is delayed implementation of caching, but that's a small risk given how it self-prioritizes if it becomes necessary.
3. Implement a more generic caching solution and utilize it here (would not recommend due to scope creep)
4. Replace the caching logic with some pre-existing generic utilities such as `cache_function_result`. I'm not sure if this would ultimately be better than removing the caching entirely or not, but at least it'd reduce the code complexity overall.

My suggestion would be to go with option 2 or 4, preferably option 2 if the performance doesn't seem horrible.

Reply:
@MythicManiac 

> Tell me that the caching is necessary for the performance of this view

We haven't really discussed what our performance budget is, and I suspect you might be more strict on this matter than I am. Probably a discussion we should have at some point.

I haven't really done any proper testing, but locally fetching an uncached page is currently in the 100ms range. I've no idea if some multiplier should be used to estimate what that means in production env.

> My suggestion would be to go with option 2 or 4, preferably option 2 if the performance doesn't seem horrible.

Objectively I'd also say that 2 might be the way to go here. But! If we ditch the caching, I'd rather also ditch `CachedPaginator`, which in turn would allow me to make this view a more standard DRF list view (probably). You might recall from "dailies" that I've mentioned trying to implement this in DRFish way was something I spent time on initially, but had to discard the idea due to the custom paginator. Is all this included in what you meant by "removing the caching logic"?

If I end up doing all that, I'd prefer to do it in a separate PR, after I've finished the client side changes related to this endpoint, so my current WIP on that side doesn't rot away with the continual changes going on in that repo currently.

Reply:
> If we ditch the caching, I'd rather also ditch `CachedPaginator`

Yep assumed that to be covered under caching in my suggestion.

> I'd prefer to do it in a separate PR

I assume you're referring to refactoring the view (after caching is removed from the current version & this merged)?